ambry-store/src/main/java/com/github/ambry/store/CompactAllPolicyFactory.java

```diff
@@ -66,20 +68,74 @@ public CompactionDetails getCompactionDetails(long totalCapacity, long usedCapac
         String sizeLog = "";
         if (validDataSizeByLogSegment != null) {
           sizeLog = blobStoreStats.dumpLogSegmentSize(validDataSizeByLogSegment.getSecond(), segmentCapacity, dataDir);
         }
+        //@formatter:off
+        List<LogSegmentName> finalLogSegmentNames =
+            storeConfig.storeCompactAllPolicyFilterOutAllValidSegment && validDataSizeByLogSegment != null
+            ?filterOutAllValidLogSegment(logSegmentsNotInJournal, blobStoreStats.getLogSegmentSizeProvider(), validDataSizeByLogSegment.getSecond())
+            :logSegmentsNotInJournal;
+        //@formatter:on
+        if (finalLogSegmentNames == null || finalLogSegmentNames.isEmpty()) {
+          return null;
+        }
+        details =
+            new CompactionDetails(time.milliseconds() - messageRetentionTimeInMs, finalLogSegmentNames, null, true);
+
         logger.info("[CAPACITY] Generating CompactionDetails {} using CompactAllPolicy for {}. Stats {}", details,
             dataDir, sizeLog);
       }
     }
     return details;
   }
+
+  /**
+   * Filter out all the leading and trailing log segments whose data is 100% valid.
+   * For instance, if segment_1, segment_2, segment_3 all have 100% valid data, this method would return a null.
+   * I segment_1, segment_3 has 100% valid data, this method would return segment_2.
+   * @param logSegmentsNotInJournal The list of {@link LogSegmentName}s that are not in journal.
+   * @param sizeProvider The {@link LogSegmentSizeProvider}.
+   * @param validDataSizeByLogSegment The map from {@link LogSegmentName} to valid data size.
+   * @return A list of {@link LogSegmentName}s.
+   */
+  List<LogSegmentName> filterOutAllValidLogSegment(List<LogSegmentName> logSegmentsNotInJournal,
+      LogSegmentSizeProvider sizeProvider, NavigableMap<LogSegmentName, Long> validDataSizeByLogSegment) {
+    List<LogSegmentName> allValidLogSegments = new ArrayList<>();
+
+    Predicate<LogSegmentName> isLogSegmentDataAllValid = segmentName -> {
+      long segmentTotalSize = sizeProvider.getLogSegmentSize(segmentName);
+      long segmentValidSize = validDataSizeByLogSegment.getOrDefault(segmentName, 0L);
+      if (segmentValidSize == segmentTotalSize) {
+        logger.info("All LogSegment {}'s data is valid, total size is {}, skip this log segment in compaction",
+            segmentName, segmentTotalSize);
+        allValidLogSegments.add(segmentName);
+        return true;
+      }
+      return false;
+    };
+    for (LogSegmentName segmentName : logSegmentsNotInJournal) {
+      if (!isLogSegmentDataAllValid.test(segmentName)) {
+        break;
+      }
+    }
+    if (allValidLogSegments.size() == logSegmentsNotInJournal.size()) {
+      logger.info("All LogSegments have 100% of valid data, skip compaction");
+      return null;
+    }
+
+    for (int i = logSegmentsNotInJournal.size() - 1; i >= 0; i--) {
```

Comment:
I think we are filtering the top consecutive 100% valid segments and bottom consecutive valid segments.
So here, shall we do 
for (int i = logSegmentsNotInJournal.size() - 1; i >= allValidLogSegments.size(); i--) {

Reply:
if we are here, then it means there is a log segment that is not 100% valid in the middle of this list, which means we would break this for loop before reaching the log segment that is last element of `allValidLogSegments`.

Reply:
I think both
 for (int i = logSegmentsNotInJournal.size() - 1; i >= 0; i--) {
and
 for (int i = logSegmentsNotInJournal.size() - 1; i >= allValidLogSegments.size(); i--) {
Will work.

The second one probably is easier to read. The while loop will break before allValidLogSegments.size.

Reply:
updated